datadog/resource_datadog_integration_slack.go

```diff
@@ -0,0 +1,220 @@
+				Description: "Slack channels to use.",
+				Type:        schema.TypeSet,
+				Required:    true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"channel_name": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"transfer_all_user_comments": {
+							Type:     schema.TypeBool,
+							Optional: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func buildIntegrationSlack(d *schema.ResourceData, client *datadog.Client) (*datadog.IntegrationSlackRequest, bool, error) {
+	slackActualIntegration, err := client.GetIntegrationSlack()
+	slackRequest := &datadog.IntegrationSlackRequest{}
+
+	needToBeCreate := false
+	if err != nil {
+		// This is a real error
+		if !strings.Contains(err.Error(), "slack not found") {
+			return nil, false, err
+
+		}
+		// slack integration is not created right now
+		needToBeCreate = true
+	}
+
+	if slackActualIntegration != nil {
+		// add existing channels except channels link to the wanted hook
+		for _, channel := range slackActualIntegration.Channels {
+			if channel.GetAccount() == d.Get("account").(string) {
+				continue
+			}
+			slackRequest.Channels = append(slackRequest.Channels, channel)
+		}
+	}
+
+	serviceHook := datadog.ServiceHookSlackRequest{}
+	// add service hook
+	serviceHook.SetUrl(d.Get("url").(string))
+	serviceHook.SetAccount(d.Get("account").(string))
+	serviceHooks := []datadog.ServiceHookSlackRequest{
+		serviceHook,
+	}
+
+	slackRequest.ServiceHooks = serviceHooks
+
+	// add channels
+	for _, channelI := range d.Get("channels").(*schema.Set).List() {
+		channel := channelI.(map[string]interface{})
+		slackChannel := datadog.ChannelSlackRequest{}
+		slackChannel.SetAccount(d.Get("account").(string))
+		slackChannel.SetChannelName(channel["channel_name"].(string))
+		slackChannel.SetTransferAllUserComments(channel["transfer_all_user_comments"].(bool))
+		slackRequest.Channels = append(slackRequest.Channels, slackChannel)
+	}
+
+	return slackRequest, needToBeCreate, nil
+}
+
+func resourceDatadogIntegrationSlackCreate(d *schema.ResourceData, meta interface{}) error {
+	integrationSlackMutex.Lock()
+	defer integrationSlackMutex.Unlock()
+
+	client := meta.(*datadog.Client)
+	slack, needCreation, err := buildIntegrationSlack(d, client)
+	if err != nil {
+		return err
+	}
+
+	if needCreation {
+		if err := client.CreateIntegrationSlack(slack); err != nil {
+			return fmt.Errorf("Failed to create integration slack using Datadog API: %s", err.Error())
+		}
+	} else {
+		if err := client.UpdateIntegrationSlack(slack); err != nil {
+			return fmt.Errorf("Failed to update integration slack using Datadog API: %s", err.Error())
+		}
+	}
+	d.SetId(d.Get("account").(string))
+
+	return nil
+}
+
+func resourceDatadogIntegrationSlackRead(d *schema.ResourceData, meta interface{}) error {
+	integrationSlackMutex.Lock()
+	defer integrationSlackMutex.Unlock()
+
+	client := meta.(*datadog.Client)
+	account := d.Id()
+	d.SetId("")
```

Comment:
why do you set it to `""` if [later](https://github.com/terraform-providers/terraform-provider-datadog/pull/411/files#diff-e35c3d1575be0cb0b644b1b7caaac061R146) you check for its value?

Reply:
I set it to blank in a first place and re-set it if hook was found. I updated code to be more readable using a boolean var instead.